[jit] add docs for serialization

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23457 [jit] Move overview to docs/ folder
* **#23456 [jit] add docs for serialization**

Differential Revision: [D16552602](https://our.internmc.facebook.com/intern/diff/D16552602)

suo committed:
[jit] add docs for serialization

suo committed:
Update on "[jit] add docs for serialization"

suo committed:
Update on "[jit] add docs for serialization"

suo commented:
ping @ZolotukhinM, you mentioned you had some questions about this :)

suo committed:
Update on "[jit] add docs for serialization"


Differential Revision: [D16552602](https://our.internmc.facebook.com/intern/diff/D16552602)

suo committed:
Update on "[jit] add docs for serialization"


Differential Revision: [D16552602](https://our.internmc.facebook.com/intern/diff/D16552602)

ZolotukhinM commented on torch/csrc/jit/docs/serialization.md:
```diff
@@ -0,0 +1,237 @@
+# Torchscript serialization
+
+This document explains the Torchscript serialization format, and the anatomy of a call to `torch::jit::save()` or `torch::jit::load()`.
```
s/Torchscript/TorchScript/

ZolotukhinM commented on torch/csrc/jit/docs/serialization.md:
```diff
@@ -0,0 +1,237 @@
+
+You'll notice that there are `.py` and `.pkl` files in this archive. That's because our serialization format tries to mimic Python's. All "code-like" information (methods, modules, classes, functions) are stored as human-readable `.py` containing valid Python syntax, and all "data-like" information (attributes, objects, etc.) are pickled using a subset of Python's pickle protocol.
+
+A model is really a top-level module with some submodules, parameters, and so on depending on what the author needs. So, `data.pkl` contains the pickled top-level module. Deserializing the model is as simple as calling `unpickle()` on `data.pkl`, which will restore the module object with its associated code and data.
```
> Deserializing the model is as simple as calling `unpickle()` on `data.pkl`

What confuses me in this paragraph is that you were previously saying that pickle is only for data - how do we pick up the code as well from this? Some clarification might be helpful.

suo replied:
Thanks for the feedback! I will add a section on how models are loaded, hopefully that will clarify these questions

ZolotukhinM reviewed (approved):
Super-helpful doc, thank you!

Krovatkin reviewed (approved):
Added a few comments.

suo committed:
Update on "[jit] add docs for serialization"


Differential Revision: [D16552602](https://our.internmc.facebook.com/intern/diff/D16552602)
